src/tools/rbd_nbd/rbd-nbd.cc

```diff
@@ -354,7 +354,6 @@ class NBDServer
     void* entry() override
     {
       (server.*func)();
-      server.shutdown();
```

Comment:
Without this unmap does not work, because  writer thread never learns that it needs to shut down. Either return this or better add  `shutdown()` in `reader_entry` after "disconnect request received".

BTW, are you testing your changes before pushing?